consensus/istanbul/core/backlog.go

```diff
@@ -103,114 +116,181 @@ func (c *core) storeBacklog(msg *istanbul.Message, src istanbul.Validator) {
-			continue
-		}
+	for _, seq := range c.getSortedBacklogSeqs() {
 
-		logger := c.logger.New("from", src, "state", c.state, "cur_round", c.current.Round(), "cur_seq", c.current.Sequence(), "func", "processBacklog")
-		isFuture := false
-
-		// We stop processing if
-		//   1. backlog is empty
-		//   2. The first message in queue is a future message
-		for !(backlog.Empty() || isFuture) {
-			m, prio := backlog.Pop()
-			msg := m.(*istanbul.Message)
-			var view *istanbul.View
-			switch msg.Code {
-			case istanbul.MsgPreprepare:
-				var m *istanbul.Preprepare
-				err := msg.Decode(&m)
-				if err == nil {
-					view = m.View
-				}
-			case istanbul.MsgPrepare:
-				fallthrough
-			case istanbul.MsgCommit:
-				var sub *istanbul.Subject
-				err := msg.Decode(&sub)
-				if err == nil {
-					view = sub.View
+		logger := c.logger.New("state", c.state, "seq", seq)
+
+		if seq < c.currentView().Sequence.Uint64() {
+			// Earlier sequence. Prune all messages.
+			c.drainBacklogForSeq(seq, nil)
+		} else if seq == c.currentView().Sequence.Uint64() {
+			// Current sequence. Process all in order.
+			c.drainBacklogForSeq(seq, func(msg *istanbul.Message, src istanbul.Validator) {
+				var view *istanbul.View
+				switch msg.Code {
+				case istanbul.MsgPreprepare:
+					var m *istanbul.Preprepare
+					err := msg.Decode(&m)
+					if err == nil {
+						view = m.View
+					}
+				case istanbul.MsgPrepare:
+					fallthrough
+				case istanbul.MsgCommit:
+					var sub *istanbul.Subject
+					err := msg.Decode(&sub)
+					if err == nil {
+						view = sub.View
+					}
+				case istanbul.MsgRoundChange:
+					var rc *istanbul.RoundChange
+					err := msg.Decode(&rc)
+					if err == nil {
+						view = rc.View
+					}
 				}
-			case istanbul.MsgRoundChange:
-				var rc *istanbul.RoundChange
-				err := msg.Decode(&rc)
-				if err == nil {
-					view = rc.View
+				if view == nil {
+					logger.Debug("Nil view", "msg", msg)
+					// continue
+					return
 				}
-			}
-			if view == nil {
-				logger.Debug("Nil view", "msg", msg)
-				continue
-			}
-			// Push back if it's a future message
-			err := c.checkMessage(msg.Code, view)
-			if err != nil {
-				if err == errFutureMessage {
-					logger.Trace("Stop processing backlog", "msg", msg)
-					backlog.Push(msg, prio)
-					isFuture = true
-					break
+				err := c.checkMessage(msg.Code, view)
+				if err != nil {
+					if err == errFutureMessage {
+						// TODO(asa): Why is this unexpected? It could be for a future round...
+						logger.Warn("Unexpected future message!", "msg", msg)
+						//backlog.Push(msg, prio)
+					}
+					logger.Trace("Skip the backlog event", "msg", msg, "err", err)
+					return
 				}
-				logger.Trace("Skip the backlog event", "msg", msg, "err", err)
-				continue
-			}
-			logger.Trace("Post backlog event", "msg", msg)
+				logger.Trace("Post backlog event", "msg", msg)
 
-			go c.sendEvent(backlogEvent{
-				src: src,
-				msg: msg,
+				go c.sendEvent(backlogEvent{
```

Comment:
So this may be wrong, but I'm wondering if we have a bug here in that when we put messages on the backlog, we check that the address signing the messages is in the current validator set. However when we pull from the backlog, it seems to me that we no longer check that the address is still in the validator set (as it might be from a different epoch). If I see correctly, this means that a validator could possibly double their voting power by sending a future messages during the old epoch, change their key, and then be able to send another PREPARE/COMMIT with their new key. Admittedly, they would have to know the block of the future, but that still sounds bad?

Reply:
Nice catch. Changed the handler for `backlogEvents` to call `handleMsg` instead of `handleCheckedMessage`

Reply:
After double checking, we actually were okay here, but no problem in calling `handleMsg` so for safety let's do that

Reply:
Out of curiousity, how were we ok?

Reply:
yes let's be super sure of this!

Reply:
See below.. 

 * only messages from the current valSet are stored on the backlog
 * if a future messages was stored, but belongs to the previous epoch we should check it again